### PR TITLE
Pallet loans sustitution in development

### PR DESCRIPTION
**Pull request by lemunozm**
This PR removes `pallet_loans` from `development` and add `pallet_loans-ref`

# Tasks

- [x] `pallet-loans-ref` implements `PoolNav` to hook it to the pool
- [x] Adapt runtime APIs
- [x] Use `pallet-loans-ref` in `development`.
- [x] Remove `pallet-loans` from `development`.
- [x] Integration tests updated with the last loan interface

~~**WARNING, double-check if we need any migrations in altair:**~~
- ~~A migration to remove the current `pallet-loans` storage? I'm pretty sure we need this to avoid data collisions with the new pallet (the new pallet will have the same name, and some storages will have the same name too).~~
- ~~After add this pallet, next calls from `pallet-pool-system` to `nav()` will return 0. Is it ok for the pools?~~

### CODE REVIEW

File: runtime/development/src/lib.rs [hunk 1]
```diff
@@ -514,14 +515,14 @@ impl InstanceFilter<RuntimeCall> for ProxyType {
...
-				RuntimeCall::Loans(pallet_loans::Call::update_nav{..}) |
+				RuntimeCall::Loans(pallet_loans::Call::update_portfolio_valuation{..}) |
 				RuntimeCall::PoolSystem(pallet_pool_system::Call::close_epoch{..}) |
 				RuntimeCall::PoolSystem(pallet_pool_system::Call::submit_solution{..}) |
 				RuntimeCall::PoolSystem(pallet_pool_system::Call::execute_epoch{..}) |
 				RuntimeCall::Utility(pallet_utility::Call::batch_all{..}) |
 				RuntimeCall::Utility(pallet_utility::Call::batch{..})
 			),
-			ProxyType::Price => matches!(c, RuntimeCall::Loans(pallet_loans::Call::price { .. })),
+			ProxyType::Price => false,
```
Comment (lemunozm) [inline runtime/development/src/lib.rs:525]: Not sure what to put here. Maybe `ProxyType::Price` is no longer needed as a variant?

Comment (branan) [reply runtime/development/src/lib.rs:525]: Yeah, now that pricing is built into creation there's no separate step

Comment (lemunozm) [reply runtime/development/src/lib.rs:525]: Ok, removed in `development`. 

If I remove it too in `Altair`, do we need a migration for `pallet_proxy` that uses it?

File: pallets/loans-ref/src/lib.rs [hunk 2]
```diff
@@ -240,14 +240,14 @@ pub mod pallet {
...
-		BoundedVec<WriteOffState<T::Rate>, T::MaxWriteOffGroups>,
+		BoundedVec<WriteOffState<T::Rate>, T::MaxWriteOffPolicySize>,
 		ValueQuery,
 	>;
 
 	/// Stores the portfolio valuation associated to each pool
 	#[pallet::storage]
 	#[pallet::getter(fn portfolio_valuation)]
-	pub(crate) type LatestPortfolioValuations<T: Config> =
+	pub(crate) type LatestPortfolioValuation<T: Config> =
```
Comment (offerijns) [inline pallets/loans-ref/src/lib.rs:250]: I wonder if we should just change this to `PortfolioValuation`? A lot of storage values can change but don't include the latest prefix. E.g. `WriteOffPolicy` is also storing the latest policy.

**Review by offerijns — APPROVED** [review]

File: pallets/loans-ref/src/lib.rs (hunk 2 above)
Comment (lemunozm) [reply pallets/loans-ref/src/lib.rs:250]: It makes sense!

File: pallets/pool-system/src/benchmarking.rs [hunk 3]
```diff
@@ -50,8 +50,9 @@ benchmarks! {
 		T::AccountId: EncodeLike<<T as frame_system::Config>::AccountId>,
 		<<T as frame_system::Config>::Lookup as sp_runtime::traits::StaticLookup>::Source:
 			From<<T as frame_system::Config>::AccountId>,
-		T::NAV: PoolNAV<T::PoolId, T::Balance, RuntimeOrigin = T::RuntimeOrigin, ClassId = u64>,
+		T::NAV: PoolNAV<T::PoolId, T::Balance, RuntimeOrigin = T::RuntimeOrigin>,
 		T::Permission: Permissions<T::AccountId, Ok = ()>,
+		<T::NAV as PoolNAV<T::PoolId, T::Balance>>::ClassId: From<u16>,
```
Comment (mustermeiszer) [inline pallets/pool-system/src/benchmarking.rs:55]: Why is this needed if we don't use class id anymore? Or do we?

File: pallets/loans-ref/src/types.rs [hunk 4]
```diff
@@ -700,7 +700,7 @@ impl<T: Config> ActiveLoan<T> {
 	}
 }
 
-#[cfg(any(test, feature = "runtime-benchmarks"))]
+#[cfg(any(feature = "std", feature = "runtime-benchmarks"))]
```
Comment (mustermeiszer) [inline pallets/loans-ref/src/types.rs:703]: Why do we need test utils under standard feature? This also reaches the node build then.

File: pallets/pool-system/src/impls.rs [hunk 5]
```diff
@@ -482,8 +482,6 @@ mod benchmarks_utils {
 				pool.parameters.max_nav_age = 999_999_999_999;
 			});
 
-			T::NAV::update_nav(pool_id).unwrap();
-
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/pool-system/src/impls.rs:486]: A non existing nav should be none.

[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/pool-system/src/impls.rs:486]: There is even an issue for that IIrc

[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/pool-system/src/impls.rs:486]: Or an dispatch error if the pool does not have assets. Mmh. Can we actually check that?

[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/pool-system/src/impls.rs:486]: Nevermind. We will not be able to check that on the loan side. But it feels a little sketchy.

File: pallets/loans-ref/src/types.rs (hunk 4 above)
Comment (mustermeiszer) [reply pallets/loans-ref/src/types.rs:703]: For the pool tests? Then I would like to have a test feature flag

**Review by mustermeiszer — COMMENTED** [review]
Why do we already expose loans to Altair? 

Otherwise, have some minor comments a few things I like to have changed.

File: pallets/loans-ref/src/types.rs (hunk 4 above)
Comment (lemunozm) [reply pallets/loans-ref/src/types.rs:703]: The wasm is not built with `std`. It is used the `integration-test` to build `LoanInfo`. I've followed a similar schema as found in the system pallet for `std_block_number()` and other utilities there.

File: pallets/pool-system/src/benchmarking.rs (hunk 3 above)
Comment (lemunozm) [reply pallets/pool-system/src/benchmarking.rs:55]: I've tried to not touch `PoolNAV` trait, in order to be updated in #908. By now `ClassId` represents the `ItemId` in the `pallet-loans-ref` implementation. We need this extra bound to be able to create them in the benchmark.

File: runtime/development/src/lib.rs [hunk 6]
```diff
@@ -413,7 +413,6 @@ pub enum ProxyType {
 	_Staking, // Deprecated ProxyType, that we are keeping due to the migration
 	NonProxy,
 	Borrow,
-	Price,
```
[the remark this answers is not in the document]
Comment (lemunozm) [reply runtime/development/src/lib.rs:416]: The `Price` is only in development, not Altair, do we need migrations there?
I do not fully understand this part, so it's ok for me to deprecate it 👍🏻 

> will be carried over as a mistake to production runtimes
Why this?

File: pallets/pool-system/src/impls.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/pool-system/src/impls.rs:486]: In the previous version, an `update_nav` was mandatory because it set the nav value that later in `close_epoch` read by `nav()`. Now, as it is stored, getting the nav over an empty or non-existent pool will return 0. 

I can return a `None` if the pool does not exist if it's preferred, although it adds a new error you need to deal with.

I feel these methods are strange for belonging to the loan. Why not adding a new method to `PoolMutate` trait that allows setting the portfolio valuation in the pool? instead of being the pool that fetches the nav? I feel like the portfolio valuation value belongs to the pool.

**Comment by lemunozm** [comment]
> Why do we already expose loans to Altair?

I've supposed we need this before going to centrifuge. If not, I can remove the Altair sustitution

File: pallets/pool-system/src/impls.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/pool-system/src/impls.rs:486]: > There is even an issue for that IIrc

Can you reference this?

**Comment by mustermeiszer** [comment]
I would remove it from Altair. And only add it if we want to release there.

File: runtime/development/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply runtime/development/src/lib.rs:416]: If it did not made it yet to Altair let's do that!

File: pallets/loans-ref/src/types.rs (hunk 4 above)
Comment (mustermeiszer) [reply pallets/loans-ref/src/types.rs:703]: In theory nodes can also use native runtimes. So it can hit your application. 

Just thinking clear features here do not hurt. But won't block.

File: pallets/pool-system/src/impls.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/pool-system/src/impls.rs:486]: I like the idea. But would not add to poolmutate but rather a new trait. Just to not expose it to the registry. Maybe some consumer like logic. Something like valuation updated and then the consumer can do with it what it want.

[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/pool-system/src/impls.rs:486]: At the same time it somehow does belong to loans. I am undecided

[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/pool-system/src/impls.rs:486]: Can't find the ticket

[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/pool-system/src/impls.rs:486]: I feel like the action belongs to the loans, but the storage to the pool. So the entity in charge of dispatching the `update_nav` should be the loan and then stored in the pool. At least this is how I see it.

File: pallets/loans-ref/src/types.rs (hunk 4 above)
Comment (lemunozm) [reply pallets/loans-ref/src/types.rs:703]: I was reading about it, and it seems like your node has a native binary and the wasm binary (that is what it will be on-chain). The native binary is part of your executable but the content is not stored on-chain, so these methods only bloat the local executable. But I think that does not add damage. And very probably, there are not even finally added because there are not used in the real node (only by the `integration-test` node), as 95% of the substrate code that is there but not used.

If I still do not convince you 😛, which feature could I add there? without creating a new one.

File: pallets/pool-system/src/impls.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/pool-system/src/impls.rs:486]: By now, is it ok returning 0, or do you prefer `None` for existing pools that never have called `update_nav` before?

File: runtime/integration-tests/src/xcm/development/tests/connectors.rs [hunk 7]
```diff
@@ -250,21 +254,6 @@ fn update_token_price() {
...
-				pool_id.clone(),
-				tranche_id.clone(),
-				Domain::EVM(1284),
-			),
-			pallet_connectors::Error::<development_runtime::Runtime>::MissingTranchePrice
-		);
-
-		Loans::update_nav(RuntimeOrigin::signed(ALICE.into()), pool_id.clone())
-			.expect("Should update nav");
-
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply runtime/integration-tests/src/xcm/development/tests/connectors.rs:267]: Nice. I think we should prohibit to call this function if the pool does not exist. But that is a connectors thing.

**Review by mustermeiszer — APPROVED** [review]
LGTM! Thanks again for fast and clean action!

File: runtime/integration-tests/src/xcm/development/tests/connectors.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lemunozm) [reply runtime/integration-tests/src/xcm/development/tests/connectors.rs:267]: I take note, for the refactor on #908 👍🏻
